web/packages/teleterm/src/ui/DocumentGatewayCliClient/DocumentGatewayCliClient.tsx

```diff
@@ -0,0 +1,158 @@
+  //
+  // After the gateway is closed, the CLI client will not be able to interact with the gateway
+  // target, but the user might still want to inspect the output.
+  if (gateway || hasRenderedTerminal) {
+    if (!hasRenderedTerminal) {
+      setHasRenderedTerminal(true);
+    }
+
+    return <DocumentTerminal doc={doc} visible={visible} />;
+  }
+
+  return <WaitingForGateway doc={doc} visible={visible} />;
+};
+
+const TIMEOUT_SECONDS = 10;
+
+const WaitingForGateway = (props: {
+  doc: types.DocumentGatewayCliClient;
+  visible: boolean;
+}) => {
+  const { doc, visible } = props;
+  const ctx = useAppContext();
+  const { documentsService } = useWorkspaceContext();
+  // If we depended on doc.status for hasTimedOut instead of using a separate state, then on reopen
+  // the doc would have status set to 'connected' on 'error' and it'd be updated from useEffect,
+  // meaning that there would be a brief flash of old state.
+  const [hasTimedOut, setHasTimedOut] = useState(false);
+
+  useEffect(() => {
+    // Update the doc state to make the progress bar show up in the tab bar.
+    // Once DocumentTerminal is mounted, it is going to update the status to 'connected' or 'error'.
+    documentsService.update(doc.uri, { status: 'connecting' });
+
+    const timeoutId = setTimeout(() => {
+      setHasTimedOut(true);
+      documentsService.update(doc.uri, { status: 'error' });
+    }, TIMEOUT_SECONDS * 1000);
+
+    return () => {
+      clearTimeout(timeoutId);
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  const openConnection = () => {
+    connectToDatabase(
+      ctx,
+      {
+        uri: doc.targetUri,
+        name: doc.targetName,
+        protocol: doc.targetProtocol,
+        dbUser: doc.targetUser,
+      },
+      { origin: 'reopened_session' }
+    );
+  };
+
+  return (
+    <Document visible={visible} px={2}>
+      <WaitingForGatewayContent
+        doc={doc}
+        hasTimedOut={hasTimedOut}
+        openConnection={openConnection}
+      />
+    </Document>
+  );
+};
+
+export const WaitingForGatewayContent = ({
+  doc,
+  hasTimedOut,
+  openConnection,
+}: {
+  doc: types.DocumentGatewayCliClient;
+  hasTimedOut: boolean;
+  openConnection: () => void;
+}) => (
+  <Flex gap={4} flexDirection="column" mx="auto" alignItems="center" mt={100}>
+    {hasTimedOut ? (
+      <div>
+        <StyledText>
+          A connection to <strong>{doc.targetName}</strong> as{' '}
+          <strong>{doc.targetUser}</strong> has not been opened up within{' '}
+          {TIMEOUT_SECONDS} seconds.
+        </StyledText>
+        <StyledText>Please try to open the connection manually.</StyledText>
+      </div>
+    ) : (
+      <StyledText>
+        Waiting for a db connection to <strong>{doc.targetName}</strong> as{' '}
+        <strong>{doc.targetUser}</strong> to be opened up.
+      </StyledText>
+    )}
+
+    <ButtonPrimary onClick={openConnection}>Open the connection</ButtonPrimary>
```

Comment:
WDYT about disabling the button while waiting for a db connection? 
Or even showing it only after timeout?

IMO the current behavior with the button always active is slightly confusing.

Reply:
> WDYT about disabling the button while waiting for a db connection?

What if you know that the connection won't be opened because the tab with the gateway is not present? I'd rather let the user click the button immediately. Also, clicking this button, even repeatedly, is not going to cause any side effects besides switching to the gateway tab.

All the "waiting" here is fake after all.